### DISCUSSION TITLE
CFS GenEnsProd File Issue

### DISCUSSION THREAD

**Discussion by ShannonShields-NOAA** [dtcenter/METplus#2114] [Configuration] [answered]
Hello, I am having a problem with GenEnsProd output files for the CFS model. When I use plot_data_plane to check the gen_ens_prod.nc files, they look like the following:
For Sea Ice Concentration:
![gen_ens_prod_CFS_seaice_Concentration_FHR840_20230408_000000V_ens](https://user-images.githubusercontent.com/66634320/230984361-e05df91d-5770-4a23-aab7-4af7993fb201.png)
For SST:
![gen_ens_prod_CFS_sst_SST_FHR840_20230408_000000V_ens](https://user-images.githubusercontent.com/66634320/230984744-905d5419-e555-476e-ab96-a4917264fd37.png)
There seems to be a regridding issue perhaps? The sea ice map should look something like the following:
![GEFS_ICEC_Z0_WEEKLYAVG_init2023032900_valid2023032900to2023040500](https://user-images.githubusercontent.com/66634320/230985537-9ab73fb1-b9e0-4293-9335-778cd64406de.png)
Maybe there is an issue in my GenEnsProd config file? This config file can be found on WCOSS2 (along with raw CFS member data and gen_ens_prod files) here: /lfs/h2/emc/vpppg/noscrub/shannon.shields/cfs_issue/files/
The spatial maps shown above can be found here: /lfs/h2/emc/vpppg/noscrub/shannon.shields/cfs_issue/spatial_maps/
Thank you for the assistance!

**Comment by JohnHalleyGotway** [comment]
Hi Shannon, are you able to post these sample files to our anonymous FTP site following [these instructions](https://github.com/dtcenter/METplus/discussions/954)? Very few staff have WCOSS2 access and posting it to a more accessible place would make it easier to debug.

I'd like to see both the Gen-Ens-Prod output as well as at least one input file. For example, whatever input file you used to create the "weekly_avg_seaice_WeeklyAvg_Concentration_init ... " plot.

Thanks!

**Reply by ShannonShields-NOAA** [reply]
I was able to transfer the files to Hera at /scratch2/NCEPDEV/stmp1/Shannon.Shields/data/evs/cfs_issue. Let me know if this is accessible, otherwise I will try FTP. The "files" directory contains the raw CFS (and GEFS for comparison) member files which are read into the GenEnsProd config file which then outputs the gen_ens_prod files. The ENS_MEAN is the output variable of interest and it is what is plotted in the maps. As shared previously, something is strange with the CFS gen_ens_prod output compared to the GEFS which is shown below:
GEFS Sea Ice Concentration:
![gen_ens_prod_GEFS_seaice_Concentration_FHR840_20230408_000000V_ens](https://user-images.githubusercontent.com/66634320/231041681-5d1cfb4b-ddaa-4c2d-8bf9-f47fd56fb9f6.png)
GEFS SST:
![gen_ens_prod_GEFS_sst_SST_FHR840_20230408_000000V_ens](https://user-images.githubusercontent.com/66634320/231041729-85736571-c9af-405a-9a6b-cfd08a8d80c7.png)

**Reply by JohnHalleyGotway** [reply]
@ShannonShields-NOAA thanks for copying over those sample files to hera. I was able to grab them there. The issue is in reading data from CFS GRIB2 files themselves. Whenever getting started with new datasets in MET, its always a good idea to run them through `plot_data_plane` to see how MET is reading them.
```
plot_data_plane cfs.ens01.2023030400.f840 ~/cfs_TMP_Z2.ps 'name="TMP"; level="Z2";'
```
And here's the result:
<img width="429" alt="Screen Shot 2023-04-11 at 11 27 25 AM" src="https://user-images.githubusercontent.com/21087144/231241959-001e2d72-7513-41fc-8478-3316169bd7a1.png">

I double-checked by plotting this GRIB2 file using Panoply and the result looks fine.
<img width="1058" alt="Screen Shot 2023-04-11 at 11 36 17 AM" src="https://user-images.githubusercontent.com/21087144/231243888-90f61bb2-516f-4a1d-96f2-7bc9a70b176c.png">

That indicates to me that this is a bug in MET's reading of GRIB2 data on a Gaussian grid.

I did some more testing... converting the GRIB2 file to GRIB1 and plotting it with MET. This result looks a bit more sane:
<img width="704" alt="Screen Shot 2023-04-11 at 11 47 43 AM" src="https://user-images.githubusercontent.com/21087144/231246434-21707a7b-bece-47b6-99b7-cadd428c7db2.png">

But it's still problematic, as seen by the clear eastward shift of values in South America.

So this is a bug.

**Reply by JohnHalleyGotway** [reply]
Shannon, can you tell me more about the origin of these files:
```
cfs.ens01.2023030400.f840
cfs.ens02.2023030400.f840
cfs.ens03.2023030400.f840
cfs.ens04.2023030400.f840
```
Running them through wgrib2, I see they're defined on a Gaussian grid:
```
	Gaussian grid: (384 x 190) units 1e-06 input WE:NS output WE:SN
	number of latitudes between pole-equator=95 #points=72960
	lat 89.277000 to -89.277000
	lon 0.000000 to 359.062000 by 0.938000
```
And that appears to match [NCEP Grid 126](https://www.nco.ncep.noaa.gov/pmb/docs/on388/tableb.html#GRID126).

All the CFS GRIB2 data files I see on [NOMADS](https://nomads.ncep.noaa.gov/pub/data/nccf/com/cfs/) appear to be on a 360x181 latlon grid:
```
	lat-lon grid:(360 x 181) units 1e-06 input WE:NS output WE:SN res 48
	lat 90.000000 to -90.000000 by 1.000000
	lon 0.000000 to 359.000000 by 1.000000 #points=65160
```
Is CFS data produced and/or distributed on both Gaussian and Lat/Lon grids?

**Reply by ShannonShields-NOAA** [reply]
Thank you John for looking into this. I thought the CFS ensemble member files were on a 1 degree lat lon grid, so I have to look into this. Maybe something is going wrong in my prep data step. Mallory Row is also using these same CFS files for her EVS work, so I'll ask what she is doing with her files.

**Reply by JohnHalleyGotway** [reply, marked as answer]
Aha, wait. I understand better what's going on here! MET *assumes* that all records in a single GRIB file are all on the same grid. This CFS file violates that assumption, and that's the source of the problem. The CFS files you sent contain a mixture of records on 2 different grids:

1. A global lat/lon grid:
```
	lat-lon grid:(360 x 181) units 1e-06 input WE:NS output WE:SN res 48
	lat 90.000000 to -90.000000 by 1.000000
	lon 0.000000 to 359.000000 by 1.000000 #points=65160
```
2. A gaussian grid:
```
	Gaussian grid: (384 x 190) units 1e-06 input WE:NS output WE:SN
	number of latitudes between pole-equator=95 #points=72960
	lat 89.277000 to -89.277000
	lon 0.000000 to 359.062000 by 0.938000

```

You can see their use alternating in this wgrib2 output:
```
wgrib2 -Sec3 cfs.ens01.2023030400.f840.grib2 
1:0:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
2:56875:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
3:89955:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
4:130313:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
5:145672:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
6:158954:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
7:240656:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
8:320019:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
9:351041:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
10:383591:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
11:440178:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
12:490259:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
13:549876:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
14:606774:Sec3 len=72 src gdef=0 npts=65160 Grid Def Template=3.0 opt arg=0
15:656632:Sec3 len=72 src gdef=0 npts=72960 Grid Def Template=3.40 opt arg=0
```

MET assumes the grid of the first record applies to all records in the file. So it's using the 2m temperature Guassian data from record number 7 to populate the lat/lon grid from record number 1.

When I use wgrib2 to select out record 7 by itself, MET plots it just fine:
<img width="577" alt="Screen Shot 2023-04-11 at 12 37 23 PM" src="https://user-images.githubusercontent.com/21087144/231258001-24292221-cb72-48ba-9adf-7d0e8eca28eb.png">

Based on this assumption that MET is making for better or for worse (for worse in this case!), please step back to see how the `cfs.ens01.2023030400.f840.grib2` file is being cobbled together. And please segregate the records on different grids into different files.

**Reply by AliciaBentley-NOAA** [reply]
Hi everyone, 

I found the following information on the CPC webpage:
https://www.ftp.cpc.ncep.noaa.gov/wd51we/wgrib2/tricks.wgrib2

(32)  How do I separate a grib file by grid types?

       **Sometimes a grib file may have multiple grids and want to split the file
       by grid type.**

       You can see the grid used by the file

           wgrib2 -grid IN.grb

       You can see the number of grids (and the checksum) used by the (linux/unix)

           wgrib2 IN.grb -checksum 3 | cut -f3 -d: | sort -u

                wgrib2 IN.grb -checksum 3   will write the checksum of section 3, the grid definition section
                cut -f3 -d:                 will select field 3 which is the check sum
                sort -u                     sort and only show unique line

            bash-3.2$ wgrib2 grids.grb -checksum 3 | cut -f3 -d: | sort -u
            sec3_cksum=1966855777
            sec3_cksum=2751972329
            sec3_cksum=3045129773
            sec3_cksum=3781314504
            sec3_cksum=4009958728

            You can extract a specific grid from grids.grb by, for example,

            bash-3.2$ wgrib2 grids.grb -checksum 3 | grep "sec3_cksum=1966855777" | wgrib2 grids.grb -i -grib grid1.grb
            2:169368:d=2009060500:HGT:200 mb:330 hour fcst:std dev
            
            Perhaps this would help with saving only the lat/lon grids that we need?

**Reply by ShannonShields-NOAA** [reply]
Thank you John; you explained the problem exactly. I am combining both pgbf and flxf CFS file types which must be on different grids (unlike GEFS which explains why those files look correct). Combining the pgbf and flxf files together in my prep step is what is causing the grid problem. I'll work on separating those files once I have access to WCOSS2.

**Reply by JohnHalleyGotway** [reply]
Whew, I'm relieved to hear that we don't need to work on a bugfix for this! Thanks for confirming.

As to the underlying assumption in MET, we could take a step back to revisit the assumption of the same grid being used for all the records. To my recollection, this is the first time it's come up.

Some options to consider are...

1. Fully support a mixture of grids in a single GRIB file. This would introduce some ambiguity though. Grid-Stat presumes that all comparisons in a single run are done on the same grid. And setting `regrid.to_grid = FCST;` can be set to indicate that the forecast grid should be used. If 2 variables are processed on 2 different grids, that setting becomes ambiguous.

2. We could add more validation checks so that each time we read gridded data, we make sure the grid is what we expect. And if not, print an error message stating that all input GRIB records must be on the same grid.

@AliciaBentley-NOAA and @ShannonShields-NOAA, please let me know if you think either of these approaches should be written up as an actual MET GitHub issue.

**Reply by ShannonShields-NOAA** [reply]
I don't know at this time @AliciaBentley-NOAA . @JohnHalleyGotway If the flxf file that I am using is on a Gaussian latitude/longitude grid (/scratch2/NCEPDEV/stmp1/Shannon.Shields/scripts/s2s_data/flxf2023041000.01.2023040900.grb2), will that work in METplus? It seems like all records in that file are on the same grid (same npts):
 wgrib2 -Sec3 flxf2023041000.01.2023040900.grb2
1:0:Sec3 len=72 src gdef=0 **npts=72960** Grid Def Template=3.40 opt arg=0
2:46011:Sec3 len=72 src gdef=0 **npts=72960** Grid Def Template=3.40 opt arg=0
3:91314:Sec3 len=72 src gdef=0 **npts=72960** Grid Def Template=3.40 opt arg=0
So according to our discussion above, as long as the grid npts are all the same within the one grib2 file, it should not pose a problem. The problem occurs when these npts don't match. Am I understanding this correctly?

**Reply by AliciaBentley-NOAA** [reply]
If the answer to Shannon's question is that CFS verification can still work for us if we separate the two different CFS grids into two different files, then I do not think we will be submitting any Github issues on this topic. I'll wait to see if the Gaussian grid can work in METplus. If it can, we'll be ok. Please let us know. Thanks!

**Reply by JohnHalleyGotway** [reply]
Shannon, yes, that's correct. You shouldn't run into problems as long as all the GRIB records in the file you're using are on the same grid. And whether it is a regular lat/lon grid or a gaussian grid, either should work fine.

In all the NOAA data I've seen on NOMADS, I haven't run into any examples of a single file containing multiple grids. So I don't expect this to be a systemic problem.

Of course, it's very easy to create a "mixed grid" file like this since GRIB files can literally be concatenated together.

Please LMK if things don't go as expected.

**Reply by ShannonShields-NOAA** [reply]
@JohnHalleyGotway Fantastic! That should be fine then as I have fixed the issue in my prep step and separated the pgbf and flxf CFS member files. So the pgbf files are all on a global lat/lon grid while the flxf files are all on a Gaussian grid. Thank you so much for your help with this; I greatly appreciate it!

**Comment by ShannonShields-NOAA** [comment]
Thanks John for marking this as answered. I just wanted to close this discussion by confirming that the fixed/separated CFS files do indeed look much better and more similar to the GEFS output:
![gen_ens_prod_CFS_SST_FHR000_2023041000V_ens](https://user-images.githubusercontent.com/66634320/231846517-4dcb13fa-bd08-4689-b047-cccf2a79aea0.png)
